test/foundry/Roles.t.sol

```diff
@@ -0,0 +1,217 @@
+
+        //New testErc20 contract
+        TierV2TestContract = new ReadWriteTier();
+
+        string memory errorMessage = string(
+            abi.encodePacked(
+                "AccessControl: account ",
+                StringsUpgradeable.toHexString(alice),
+                " is missing role ",
+                vm.toString(vault.ERC20TIERER())
+            )
+        );
+        vm.expectRevert(bytes(errorMessage));
+
+        //set Tier
+        vault.setERC20Tier(address(TierV2TestContract), minTier, context, data);
+    }
+
+    function testSetERC1155TierWithoutRole() public {
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
+
+        bytes memory data = "";
+        uint8 minTier = 10;
+        uint256[] memory context = new uint256[](0);
+
+        //New testErc20 contract
+        TierV2TestContract = new ReadWriteTier();
+
+        string memory errorMessage = string(
+            abi.encodePacked(
+                "AccessControl: account ",
+                StringsUpgradeable.toHexString(alice),
+                " is missing role ",
+                vm.toString(vault.ERC1155TIERER())
+            )
+        );
+        vm.expectRevert(bytes(errorMessage));
+
+        //set Tier
+        vault.setERC1155Tier(address(TierV2TestContract), minTier, context, data);
+    }
+
+    function testSnapshotWithoutRole() public {
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
+
+        bytes memory data = "";
+
+        string memory errorMessage = string(
+            abi.encodePacked(
+                "AccessControl: account ",
+                StringsUpgradeable.toHexString(alice),
+                " is missing role ",
+                vm.toString(vault.ERC20SNAPSHOTTER())
+            )
+        );
+        vm.expectRevert(bytes(errorMessage));
+
+        //snapshot
+        vault.snapshot(data);
+    }
+
+    function testCertifyWithoutRole() public {
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
+
+        // Get the current block number
+        uint256 blockNum = block.number;
+
+        // Set up expected parameters
+        uint256 certifyUntil = block.timestamp + 1000;
+        bool forceUntil = false;
+        bytes memory data = abi.encodePacked("Certification data");
+
+        string memory errorMessage = string(
+            abi.encodePacked(
+                "AccessControl: account ",
+                StringsUpgradeable.toHexString(alice),
+                " is missing role ",
+                vm.toString(vault.CERTIFIER())
+            )
+        );
+        vm.expectRevert(bytes(errorMessage));
+
+        // Call the certify function
+        vault.certify(certifyUntil, blockNum, forceUntil, data);
+
+        vm.stopPrank();
+    }
+
+    function testConfiscateWithoutRole() public {
+        // Prank as Alice for the transaction
+        vm.startPrank(alice);
+
+        bytes memory data = abi.encodePacked("Certification data");
```

Comment:
abi encoding copies bytes to a whole new region of memory but you can just cast strings directly like `bytes("foo")`

Reply:
I'll consider that in future, but actually i fuzzed the data here